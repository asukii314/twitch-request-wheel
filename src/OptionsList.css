.partyPackList {
  display: flex;
  margin: 10px;
  flex-wrap: wrap;
  text-align: left;
}

.partyPackCard {
  background-color: darkslategrey;
  border-radius: 10px;
  margin: 10px;
  padding: 20px;
}

.partyPackName > input[type='checkbox'] {
    float: right;
}

.partyPackGameList {
    list-style-type: none;
    margin: 0px;
    padding: 0px;
}

.partyPackGameList .gameName {

}

/* Extra small devices (phones, 600px and down) */
@media only screen and (max-width: 600px) {
  .partyPackCard {
    width: 80%;
  }
}

/* Small devices (portrait tablets and large phones, 600px and up) */
@media only screen and (min-width: 600px) {
  .partyPackCard {
    width: 38%;
  }
}

/* Medium devices (landscape tablets, 768px and up) */
@media only screen and (min-width: 768px) {
  .partyPackCard {
    width: 25%;
  }
}

/* Large devices (laptops/desktops, 992px and up) */
@media only screen and (min-width: 992px) {
  .partyPackCard {
    width: 18%;
  }
}

/* Extra large devices (large laptops and desktops, 1200px and up) */
@media only screen and (min-width: 1200px) {
  .partyPackCard {
    width: 18%;
  }
}

.partyPackName {
  font-size: 16px;
  margin: 0;
}

.gameName {
  font-size: 12px;
  margin-bottom: 2px;
}

.possibleGameMatch {
  font-size: 10px;
  color: darkgrey;
}
