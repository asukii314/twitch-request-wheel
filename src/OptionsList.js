import { Component } from 'react';
import YAML from 'yaml'
import rawJackboxGameList from './JackboxGames.yaml';
import './OptionsList.css';
const fetch = require('node-fetch');

export default class OptionsList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            allowedGames: [],
            validGames: []
        };
    }

    componentDidMount(props) {
        fetch(rawJackboxGameList)
            .then(r => r.text())
            .then(text => {
                let validGames = YAML.parse(text);
                // this.setState({validGames});
                this.setAllowedGames(validGames);
            });
    }

    componentDidUpdate() {
        if (!this.state.allowedGames || this.state.validGames.length === 0 || this.state.allowedGames.length === 0) return;
        localStorage.setItem('__allowedGames', JSON.stringify(this.state.allowedGames));
    }

    getAllowedGames = function(validGames) {
        let __allowedGames = localStorage.getItem('__allowedGames');
        let allowedGames = (__allowedGames) ? JSON.parse(__allowedGames) : {};


        // let output = {};
        //
        // // return list of valid games merged with existing allowedGames
        // Object.entries(validGames).forEach(([pack, packGames]) => {
        //     // get valid games within each pack
        //     Object.keys(packGames).forEach(game => {
        //         let gameId = `${pack} ${game}`.replace(/\W/ig, '_');
        //         output[gameId] = (allowedGames[gameId])
        //             ? allowedGames[gameId]
        //             : {
        //                 id: gameId,
        //                 game,
        //                 pack,
        //                 enabled: true
        //             };
        //         output[gameId] = allowedGames[gameId];
        //     });
        // });
        // return output;

        // return list of valid games merged with existing allowedGames
        return Object.assign({},
            ...[].concat(
                ...Object.entries(validGames).map(([pack, packGames]) => {
                    // get valid games within each pack
                    return Object.assign({},
                        ...Object.keys(packGames).map(game => {
                            let gameId = `${pack} ${game}`.replace(/\W/ig, '_');
                            if (gameId in allowedGames && allowedGames[gameId].id !== undefined) {
                                return {
                                    [gameId]: allowedGames[gameId]
                                };
                            }
                            // add entry with default value
                            // console.log(`Adding ${gameId}`);
                            return {
                                [gameId]: {
                                    id: gameId,
                                    game,
                                    pack,
                                    enabled: true
                                }
                            }
                        })
                    );
                })
            )
        )
    }

    setAllowedGames = function(validGames) {
        let allowedGames = this.getAllowedGames(validGames);
        this.setState({
            allowedGames,
            validGames
        });
        localStorage.setItem('__allowedGames', JSON.stringify(allowedGames));
    }

    onCheckHandler = ({target}) => {
        let {id, checked} = target;
        this.setState((prevState) => ({
            allowedGames: {
                ...prevState.allowedGames,
                [id]: {
                    ...prevState.allowedGames[id],
                    enabled: checked
                }
            }
        }));
    }

    onGroupCheckHandler = ({target}) => {
        let {value, checked} = target;
        // console.log({value, checked});
        let packGames = Object.values(this.state.allowedGames).filter(game => game.pack === value);
        this.setState((prevState) => ({
            allowedGames: {
                ...prevState.allowedGames,
                ...Object.assign(
                    ...packGames.map(g => ({
                        [g.id]: {
                            ...g,
                            enabled: checked
                        }
                    })
                ))
            }
        }));
    }

    renderGameCard = function(props) {
        let {id, game, pack} = props;
        let {allowedGames} = this.state;
        let checked = allowedGames[id]?.enabled || false;
        return (
            <li className="gameName" key={id}>
                <input type="checkbox" id={id} name={id} value={id} checked={checked} onChange={this.onCheckHandler} />
                {' '}
                <label htmlFor={id} title={`${pack}: ${game}`}>
                    {game}
                </label>
            </li>
        );
    }

    renderPackCheckbox = function(pack) {
        let packGames = Object.values(this.state.allowedGames).filter(game => game.pack === pack);
        let isChecked = packGames.filter(g => g.enabled === true).length === packGames.length;
        return (
            <input type="checkbox" id={pack} name={pack} value={pack} checked={isChecked} onChange={this.onGroupCheckHandler} />
        );
    }

    renderPartyPack = function (props) {
        const gameList = Object.keys(props.metadata);

        let packGames = Object.values(this.state.allowedGames).filter(game => game.pack === props.partyPackName);
        let gamesEnabled = packGames.filter(g => g.enabled === true)
        let isChecked = gamesEnabled.length === packGames.length;
        let isIndeterminate = (!isChecked && gamesEnabled.length !== 0)
        return (
            <div className="partyPackCard" key={props.key}>
                <p className="partyPackName">
                    {props.partyPackName} <input type="checkbox" id={props.partyPackName} name={props.partyPackName} value={props.partyPackName} checked={isChecked} onChange={this.onGroupCheckHandler} ref={el => el && (el.indeterminate = isIndeterminate)} />
                </p>
                <ul className="partyPackGameList">
                    {gameList.map(
                        (gameName, i) => this.renderGameCard({
                            game: gameName,
                            id: `${props.partyPackName} ${gameName}`.replace(/\W/ig, '_'),
                            pack: props.partyPackName
                        })
                    )}
                </ul>
            </div>
        );
    }

    render() {
        let {validGames} = this.state;
        const partyPackList = Object.keys(validGames);

        return (
            <div className="partyPackList">
                {partyPackList.map(
                    (partyPackName, i) => this.renderPartyPack({
                        key: i,
                        partyPackName,
                        metadata: this.state.validGames[partyPackName]
                    })
                )}
            </div>
        );
    }

}
