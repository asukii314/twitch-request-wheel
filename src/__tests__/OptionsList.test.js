import {createRenderer} from 'react-test-renderer/shallow';
import fetch from 'node-fetch';
import OptionsList from '../OptionsList';
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import yamlGameList from '../JackboxGames.yaml';
import YAML from 'yaml'

jest.mock('node-fetch');

const VALID_GAMES = YAML.parse(yamlGameList);
const ALLOWED_GAMES = {
    Any_Version_Quiplash:{
        id: "Any_Version_Quiplash",
        game: "Quiplash",
        pack: "Any Version",
        enabled: true
    },
    Any_Version_Trivia_Murder_Party:{
        id: "Any_Version_Trivia_Murder_Party",
        game: "Trivia Murder Party",
        pack: "Any Version",
        enabled: true
    },
    Party_Pack_1_Drawful:{
        id: "Party_Pack_1_Drawful",
        game: "Drawful",
        pack: "Party Pack",
        enabled: false
    },
    Party_Pack_1_You_Don_t_Know_Jack:{
        id: "Party_Pack_1_You_Don_t_Know_Jack",
        game: "You Don't Know Jack",
        pack: "Party Pack",
        enabled: true
    }
};

describe('OptionsList', () => {
    describe('setAllowedGames', () => {
        test('should parse localStorage item when set', () => {
            jest.spyOn(window.localStorage.__proto__, 'getItem').mockReturnValue(JSON.stringify(ALLOWED_GAMES));
            jest.spyOn(JSON, 'parse').mockReturnValue(ALLOWED_GAMES);
            let component = new OptionsList();
            jest.spyOn(component, 'setState').mockImplementation(()=>{});

            component.state.validGames = VALID_GAMES;
            component.setAllowedGames(VALID_GAMES);

            expect(window.localStorage.__proto__.getItem).toHaveBeenCalledTimes(1);
            expect(component.setState).toHaveBeenCalledTimes(1);
        });
    });

    describe('render', () => {
        test('should render without error', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<OptionsList />);

            let component = shallowRenderer.getRenderOutput();
            // console.log('component:', JSON.stringify(component, null, 2));
            expect(component.props.className).toBe('partyPackList');
            shallowRenderer.unmount();
        });
        test('should render using validGames in state', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<OptionsList />);

            let instance = shallowRenderer.getMountedInstance();
            instance.setState({
                allowedGames: ALLOWED_GAMES,
                validGames: VALID_GAMES
            });

            let component = shallowRenderer.getRenderOutput();
            expect(component.props.className).toBe('partyPackList');
            expect(component).toMatchSnapshot();
            shallowRenderer.unmount();
        });
        test('renders component with yaml content', async () => {
            fetch.mockImplementation(() => {
                return Promise.resolve({
                    text: () => Promise.resolve(yamlGameList),
                })
            });

            let props = {};
            let {container} = render(<OptionsList {...props} />);

            await waitFor(() => screen.getByText(/Trivia Murder Party/i));
            expect(screen.getByText(/Trivia Murder Party/i)).toBeInTheDocument();
            expect(container).toMatchSnapshot(); // TODO: fix incorrect checkbox values
        });
    });
});
