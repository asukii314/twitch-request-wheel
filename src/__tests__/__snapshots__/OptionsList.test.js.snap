// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OptionsList render renders component with yaml content 1`] = `
<div>
  <div
    class="partyPackList"
  >
    <div
      class="partyPackCard"
    >
      <p
        class="partyPackName"
      >
        Any Version
         
        <input
          checked=""
          id="Any Version"
          name="Any Version"
          type="checkbox"
          value="Any Version"
        />
      </p>
      <ul
        class="partyPackGameList"
      >
        <li
          class="gameName"
        >
          <input
            checked=""
            id="Any_Version_Quiplash"
            name="Any_Version_Quiplash"
            type="checkbox"
            value="Any_Version_Quiplash"
          />
           
          <label
            for="Any_Version_Quiplash"
            title="Any Version: Quiplash"
          >
            Quiplash
          </label>
        </li>
        <li
          class="gameName"
        >
          <input
            checked=""
            id="Any_Version_Trivia_Murder_Party"
            name="Any_Version_Trivia_Murder_Party"
            type="checkbox"
            value="Any_Version_Trivia_Murder_Party"
          />
           
          <label
            for="Any_Version_Trivia_Murder_Party"
            title="Any Version: Trivia Murder Party"
          >
            Trivia Murder Party
          </label>
        </li>
      </ul>
    </div>
    <div
      class="partyPackCard"
    >
      <p
        class="partyPackName"
      >
        Party Pack 1
         
        <input
          checked=""
          id="Party Pack 1"
          name="Party Pack 1"
          type="checkbox"
          value="Party Pack 1"
        />
      </p>
      <ul
        class="partyPackGameList"
      >
        <li
          class="gameName"
        >
          <input
            checked=""
            id="Party_Pack_1_You_Don_t_Know_Jack"
            name="Party_Pack_1_You_Don_t_Know_Jack"
            type="checkbox"
            value="Party_Pack_1_You_Don_t_Know_Jack"
          />
           
          <label
            for="Party_Pack_1_You_Don_t_Know_Jack"
            title="Party Pack 1: You Don't Know Jack"
          >
            You Don't Know Jack
          </label>
        </li>
        <li
          class="gameName"
        >
          <input
            checked=""
            id="Party_Pack_1_Drawful"
            name="Party_Pack_1_Drawful"
            type="checkbox"
            value="Party_Pack_1_Drawful"
          />
           
          <label
            for="Party_Pack_1_Drawful"
            title="Party Pack 1: Drawful"
          >
            Drawful
          </label>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`OptionsList render should render using validGames in state 1`] = `
<div
  className="partyPackList"
>
  <div
    className="partyPackCard"
  >
    <p
      className="partyPackName"
    >
      Any Version
       
      <input
        checked={true}
        id="Any Version"
        name="Any Version"
        onChange={[Function]}
        type="checkbox"
        value="Any Version"
      />
    </p>
    <ul
      className="partyPackGameList"
    >
      <li
        className="gameName"
      >
        <input
          checked={true}
          id="Any_Version_Quiplash"
          name="Any_Version_Quiplash"
          onChange={[Function]}
          type="checkbox"
          value="Any_Version_Quiplash"
        />
         
        <label
          htmlFor="Any_Version_Quiplash"
          title="Any Version: Quiplash"
        >
          Quiplash
        </label>
      </li>
      <li
        className="gameName"
      >
        <input
          checked={true}
          id="Any_Version_Trivia_Murder_Party"
          name="Any_Version_Trivia_Murder_Party"
          onChange={[Function]}
          type="checkbox"
          value="Any_Version_Trivia_Murder_Party"
        />
         
        <label
          htmlFor="Any_Version_Trivia_Murder_Party"
          title="Any Version: Trivia Murder Party"
        >
          Trivia Murder Party
        </label>
      </li>
    </ul>
  </div>
  <div
    className="partyPackCard"
  >
    <p
      className="partyPackName"
    >
      Party Pack 1
       
      <input
        checked={true}
        id="Party Pack 1"
        name="Party Pack 1"
        onChange={[Function]}
        type="checkbox"
        value="Party Pack 1"
      />
    </p>
    <ul
      className="partyPackGameList"
    >
      <li
        className="gameName"
      >
        <input
          checked={true}
          id="Party_Pack_1_You_Don_t_Know_Jack"
          name="Party_Pack_1_You_Don_t_Know_Jack"
          onChange={[Function]}
          type="checkbox"
          value="Party_Pack_1_You_Don_t_Know_Jack"
        />
         
        <label
          htmlFor="Party_Pack_1_You_Don_t_Know_Jack"
          title="Party Pack 1: You Don't Know Jack"
        >
          You Don't Know Jack
        </label>
      </li>
      <li
        className="gameName"
      >
        <input
          checked={false}
          id="Party_Pack_1_Drawful"
          name="Party_Pack_1_Drawful"
          onChange={[Function]}
          type="checkbox"
          value="Party_Pack_1_Drawful"
        />
         
        <label
          htmlFor="Party_Pack_1_Drawful"
          title="Party Pack 1: Drawful"
        >
          Drawful
        </label>
      </li>
    </ul>
  </div>
</div>
`;
